lib/src/component_declaration/redux_component.dart

```diff
@@ -47,117 +51,68 @@ abstract class ReduxUiProps<StoreT> extends UiProps {
-    _tearDownSubs();
+    _tearDownSub();
   }
 }
 
-/// Builds on top of [UiStatefulComponent], adding `built_redux` integration.
-///
-/// * Redux components are responsible for rendering application views and turning
-///   user interactions and events into [Action]s.
-/// * Redux components can use data from one or many [Store] instances to define
-///   the resulting component.
-///
-/// Use with the over_react transformer via the `@Component()` ([annotations.Component]) annotation.
-///
-/// > Related: [ReduxUiComponent]
-abstract class ReduxUiStatefulComponent<T extends ReduxUiProps, S extends UiState> extends UiStatefulComponent<T, S>
-    with _ReduxComponentMixin<T> {
-  @mustCallSuper
-  @override
-  void componentWillMount() {
-    super.componentWillMount();
-    _setUpSubs();
-  }
+abstract class _ReduxComponentMixin<
+    V extends Built<V, B>,
+    B extends Builder<V, B>,
+    A extends ReduxActions,
+    T extends ReduxUiProps<V, B, A>,
+    Substate> implements UiComponent<T> {
+  Substate _connectedProps;
+  Substate get connectedProps => _connectedProps;
 
-  @mustCallSuper
-  @override
-  void componentWillReceiveProps(Map nextProps) {
-    super.componentWillReceiveProps(nextProps);
-    // Unbind store subs so they can be re-bound in componentDidUpdate
-    // once the new props are available, ensuring the values returned by [redrawOn]
-    // are not outdated.
-    _tearDownSubs();
-  }
+  // implemented by the concrete component to subscribe to changes
+  // to the values from the redux store that this component cares about
+  Substate connect(covariant dynamic state) => state;
 
-  @mustCallSuper
-  @override
-  void componentDidUpdate(Map prevProps, Map prevState) {
-    // If the subs are not bound at this point, then that means they were unbound by
-    // componentWillReceiveProps, and need to be re-bound now that new props are available.
-    if (!_areStoreSubsBound) _setUpSubs();
-  }
-
-  @mustCallSuper
-  @override
-  void componentWillUnmount() {
-    super.componentWillUnmount();
-    _tearDownSubs();
-  }
-}
+  StreamSubscription _storeSub;
 
-abstract class _ReduxComponentMixin<T extends ReduxUiProps> implements UiComponent<T> {
-  List<Stream> redrawOn() {
-    if (props.store is Store) {
-      return <Stream>[props.store.stream];
-    } else {
-      return <Stream>[];
+  void _setUpSub() {
+    if (_storeSub != null) {
+      throw new StateError('Store subscription is already set up');
     }
-  }
 
-  bool get _areStoreSubsBound => _storeSubs != null;
-
-  List<StreamSubscription> _storeSubs;
-
-  void _setUpSubs() {
-    if (_areStoreSubsBound) {
-      throw new StateError('Store subscriptions are already bound.');
-    }
-
-    _storeSubs = [];
-
-    for (var stream in redrawOn()) {
-      _storeSubs.add(stream.listen((_) => redraw()));
-    }
+    _storeSub = props.store.nextSubstate(connect).listen((Substate s) {
+      _connectedProps = s;
+      redraw();
+    });
```

Comment:
I think the only problem with this is we can only listen to one substate so we'd be stuck if we wanted `state.count` and `state.todos`.

Reply:
Thoughts, @davidmarne-wf?

Reply:
No, your substate would be a built_value object like shown in the doc comment https://github.com/Workiva/over_react/pull/118/files/20dc99def5af7766e5045dbcc909ad3bb39db8f1#diff-cd6ab537d255abbf1ec247df0369120eR119

Reply:
i am of the opinion you define a new built_value and map state to that. so say your state is
```dart
abstract class State extends Built<State, StateBuilder> {
  int get count;
  BuiltList<Todo> todos;
  Map<String, String> options;
 ...
}
```
you define a model
```dart
abstract class SubstateMyComponentWants extends Built<SubstateMyComponentWants, SubstateMyComponentWantsBuilder> {
  int get count;
  BuiltList<Todo> todos;
 ...
}
```

then your connect is 
```dart
@override
SubstateMyComponentWants connect(State s) => new SubstateMyComponentWants((builder) => builder
  ..count = state.count
  ..todos = state.todos);
```

Reply:
I'ld be open to suggestions for multiple connectors though. I imagine it would look something like:
```dart
typedef Substate StateMapper<State, Substate>(State s);

Iterable<StateMapper<State, dynamic>> get connectors;

...

void _setUpSub() {
    _storeSubs.addAll(connectors.map((connect) => props.store.nextSubstate(connect).listen((_) {
      redraw();
    }));
  }
```

Reply:
then you wouldn't have connectedState, you would prob just expose a store.state getter and let the renderers pull data from that.

Reply:
or even simpler you could just have 

```dart
bool shouldUpdateOnReduxStateChange(State s);
```

that the consumer implements.

then if i wanna update on todos or count changes i do:

```dart
@override
bool shouldUpdateOnReduxStateChange(State prev, State next) =>  prev.todos != next.todos || prev.count != next.count
```

The upside to keeping how it is now, where you load everything into a new built_value, is this comparison is generated for the model

Reply:
I like the first option, with using built_value objects to contain multiple state values

Reply:
That's fine.